### DISCUSSION TITLE
How do named directories '~' work in ble.sh?

### DISCUSSION THREAD

**Discussion by mozirilla213** [akinomyoga/ble.sh#299] [Q&A] [answered]
I noticed that apparently ble.sh supports static named directories like zsh. In zsh:
```
$ hash -d b=/boot
$ cd ~b
$ pwd
/boot
```
And there's tab completion.
In ble.sh when I do
```
$ cd ~<TAB>
```
the default named directories show and I can navigate to them like in zsh, but how do I add and delete them? I don't find documentation for them.

**Comment by akinomyoga** [comment]
This isn't the feature of `ble.sh`, but the feature of Bash called the tilde expansion. There is no way to add new directories in the tilde expansions or to customize the behavior of the tilde expansion. The description is found in [3.5.2 Tilde expansion - Bash Reference Manual](https://www.gnu.org/software/bash/manual/bash.html#Tilde-Expansion). If you would like to supplement the completion candidates, you can write a programmable completion setting, though it would be expanded in place so is different from Zsh's named directories.

**Reply by mozirilla213** [reply]
Oh yeah, I didn't realize this at first, my system has a lot of users with home directories all over the place so that confused me

**Comment by akinomyoga** [comment, marked as answer]
It is not possible to change the behavior of Bash's tilde expansion, but I'm currently thinking about the approximation by sabbrev expansion with the following setup:

```bash
# blerc

# Edit (2023-03-09): Now the following two lines are already contained in ble.sh
# ble/widget/magic-insert() { ble/complete/sabbrev/expand; ble/widget/self-insert; }
# ble-bind -f / magic-insert

# The list of "named directories"
ble-sabbrev \~mydir1=/home/murase/mydir1
ble-sabbrev \~mydir2=/home/murase/mydir1/mydir2
```

Then

```bash
$ cd ~mydir1 # <-- press SP or / will expand ~mydir1 to the registered sabbrev contents
$ cd /home/murase/mydir1/ # Example
```

**Comment by mozirilla213** [comment]
Oh awesome, I immediately thought this behaviour could be replicated pretty easily somehow and yeah it can. Great thank you.

**Reply by akinomyoga** [reply]
In commit e6b9581c276c9cf5098725c5445417a4f4e28fa9, I added a new widget `magic-slash` and the default keybinding of `/` to `magic-slash`. Now the sabbrev expansions named `~*` are subject to expansions when a slash is inserted by the user.
